chirp/drivers/mml_jc8810.py

```diff
@@ -0,0 +1,1050 @@
+
+SETTING_LISTS = {
+    "abr": ABR_LIST,
+    "almod": ALMODE_LIST,
+    "autolk": AUTOLK_LIST,
+    "dtmfoff": DTMFSPEED_LIST,
+    "dtmfon": DTMFSPEED_LIST,
+    "dtmfst": DTMFST_LIST,
+    "dualtx": DUALTX_LIST,
+    "encrypt": ENCRYPT_LIST,
+    "language": LANGUAGE_LIST,
+    "mdfa": MDF_LIST,
+    "mdfb": MDF_LIST,
+    "menuquit": MENUQUIT_LIST,
+    "ponmsg": PONMSG_LIST,
+    "pttid": PTTID_LIST,
+    "pttlt": PTTLT_LIST,
+    "qtsave": QTSAVE_LIST,
+    "rpste": RPSTE_LIST,
+    "rptrl": RPSTE_LIST,
+    "rxendtail": TONERXEND_LIST,
+    "save": SAVE_LIST,
+    "scode": PTTIDCODE_LIST,
+    "screv": SCREV_LIST,
+    "tailcode": TAILCODE_LIST,
+    "tone": TONE_LIST,
+    "tot": TOT_LIST,
+    "vox": OFF1TO9_LIST,
+    "voxd": VOXD_LIST,
+    "workmode": WORKMODE_LIST
+    }
+
+
+def _enter_programming_mode(radio):
+    serial = radio.pipe
+
+    exito = False
+    for i in range(0, 5):
+        serial.write(radio._magic)
+        ack = serial.read(1)
+
+        try:
+            if ack == CMD_ACK:
+                exito = True
+                break
+        except Exception:
+            LOG.debug("Attempt #%s, failed, trying again" % i)
+            pass
+
+    # check if we had EXITO
+    if exito is False:
+        msg = "The radio did not accept program mode after five tries.\n"
+        msg += "Check you interface cable and power cycle your radio."
+        raise errors.RadioError(msg)
+
+    try:
+        serial.write(b"F")
+        ident = serial.read(8)
+    except Exception:
+        raise errors.RadioError("Error communicating with radio")
+
+    if ident not in radio._fingerprint:
+        LOG.debug(util.hexprint(ident))
+        raise errors.RadioError("Radio returned unknown identification string")
+
+
+def _exit_programming_mode(radio):
+    serial = radio.pipe
+    try:
+        serial.write(b"E")
+    except Exception:
+        raise errors.RadioError("Radio refused to exit programming mode")
+
+
+def _read_block(radio, block_addr, block_size):
+    serial = radio.pipe
+
+    cmd = struct.pack(">cHb", b'R', block_addr, block_size)
+    expectedresponse = b"R" + cmd[1:]
+    LOG.debug("Reading block %04x..." % (block_addr))
+
+    try:
+        serial.write(cmd)
+        response = serial.read(4 + block_size)
+        if response[:4] != expectedresponse:
+            raise Exception("Error reading block %04x." % (block_addr))
+
+        block_data = response[4:]
+    except Exception:
+        raise errors.RadioError("Failed to read block at %04x" % block_addr)
+
+    return block_data
+
+
+def _write_block(radio, block_addr, block_size):
+    serial = radio.pipe
+
+    cmd = struct.pack(">cHb", b'W', block_addr, block_size)
+    data = radio.get_mmap()[block_addr:block_addr + block_size]
+
+    LOG.debug("Writing Data:")
+    LOG.debug(util.hexprint(cmd + data))
+
+    try:
+        serial.write(cmd + data)
+        if serial.read(1) != CMD_ACK:
+            raise Exception("No ACK")
```

Comment:
Not a big deal but you'll never see this because it will get swallowed and raised as a radioerror. You could `LOG.error()` in the handler below if you want to see it.